posts/navarroaxel/14.public-key-with-openid-connect-jkws.md

```diff
@@ -0,0 +1,127 @@
+---
+title: Validate an OpenID Connect JWT using a public key in JWKS
+published: true
+description: How to validate a JSON web token using the public keys of the issuer
+tags: 'javascript, node, security, webdev'
+cover_image: ./assets/oidc-jwks-cover.jpg
+id: 1339462
+---
+
+You may have used OpenID Connect in the Front-end, where an IDP (IDentity Provider) authenticates a user, gives you a bunch of tokens in the browser and then you can add the `Authorization` header to your HTTP requests to your own Back-End because you trust this IDP. But what happens when you can't find where you can verify the `id_token` (or `access_token`) using some endpoint in the IDP?
+
+Well, I found how an [OpenID Connect `id_token` should be validated](https://curity.io/resources/learn/validating-an-id-token/). It wasn't straightforward in my case: I had to do a lot of research to validate my `id_token`. Let's check how to make this easy using Node.js and the [`jsonwebtoken`](https://github.com/auth0/node-jsonwebtoken) npm package made by [Auth0](https://auth0.com).
+
+## Understanding JWT to know how to validate it
+
+A JSON Web Token (JWT) is a string built with 2 JSON objects encoded in `base64` and a signature; these parts are joined by a period (`.`) with the following structure: `<header>.<payload>.<signature>`.
+
+💡 This is why a JWT always starts with `ey`, because it is the result of encoding to `base64` this: `{"`, the beginning of any JSON.
+
+In the `header` part we can find which signature algorithm was used in the `alg` attribute (e.g. _RS256_) to sign the JWT, and the `kid` attribute tells which _Key ID_ from the JSON Web Key Set (JWKS) was used for a given token.
+
+🧠 Remember that when the JWT `header` has a _Key ID_ (`kid`), JWKS is used.
+
+And here is where the problem starts. Where do I find the JWKS to get the public key so we can  verify the integrity of this token? 😫
+
+## The issuer
+
+The issuer is the one who created and signed the JWT, and we can know this by checking the value `iss` in the payload of our JWT - using `jwt.decode` from `jsonwebtoken`.
+
+```javascript
+const jwt = require('jsonwebtoken');
+const payload = jwt.decode('<your_jwt_here>');
+console.log(payload.iss);
+```
+
+Alternatively, you can paste your JWT into <https://jwt.io> (don't worry, it's a safe website from Auth0), and `iss` is there too!
+
+> In OpenID Connect, the issuer should be a URL, but could be just the name of the IDP. In that case you should read more docs to find where the JWKS URI is. 😭
+
+For our example  we can use <https://sandrino.auth0.com> as issuer, so you can know the OpenID configuration using the well-know URI <https://sandrino.auth0.com/.well-known/openid-configuration>, and in the `jwks_uri` attribute is where you can find the JWKS for our issuer. You can check this same value in another  issuer <https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration>, the `jwks_uri` is there too! 😉
```

Comment:
lo mismo en "example  we". Y en "another  issuer".

Reply:
```suggestion
For our example we can use <https://sandrino.auth0.com> as issuer, so you can know the OpenID configuration using the well-know URI <https://sandrino.auth0.com/.well-known/openid-configuration>, and in the `jwks_uri` attribute is where you can find the JWKS for our issuer. You can check this same value in another issuer <https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration>, the `jwks_uri` is there too! 😉
```